src/Proto3/Suite/DotProto/Generate.hs

```diff
@@ -880,40 +943,60 @@ fromJSONPBMessageInstD _ctxt parentIdent msgIdent messageParts = do
   msgName    <- nestedTypeName parentIdent =<< dpIdentUnqualName msgIdent
   qualFields <- getQualifiedFields msgName messageParts
 
-  -- E.g., for message Something{ oneof name_or_id { string name = _; int32 someid = _; } }:
-  -- Hs.msum
-  --   [ Just . SomethingPickOneName   <$> (HsJSONPB.parseField obj "name")
-  --   , Just . SomethingPickOneSomeid <$> (HsJSONPB.parseField obj "someid")
-  --   , pure Nothing
-  --   ]
-  let oneofParserE fld =
-        HsApp msumE (HsList ((subParserEs <> fallThruE) fld))
+  let lambdaPVar = patVar "obj"
+  let lambdaVar  = HsVar (unqual_ "obj")
+  let oneofParserE (OneofField oneofType fields) = objToParser
         where
-          fallThruE OneofField{}
-            = [ HsApp pureE (HsVar (haskellName "Nothing")) ]
-          subParserEs OneofField{subfields}
-            = subParserE <$> subfields
-          subParserE OneofSubfield{subfieldConsName, subfieldName}
-            = HsInfixApp (HsInfixApp
-                            (HsVar (haskellName "Just"))
-                            composeOp
-                            (HsVar (unqual_ subfieldConsName)))
-                         fmapOp
-                         (apply (HsVar (jsonpbName "parseField"))
-                                [ HsVar (unqual_ "obj")
-                                , HsLit (HsString (coerce subfieldName))
-                                ])
+          oneofTyLit = HsLit (HsString oneofType)
+          objToParser = HsParen $
+              HsLet [ HsFunBind [ match_ (HsIdent bndNameStr) []
+                                         (HsUnGuardedRhs tryParseDisjunctsE) []
+                                ]
+                    ]
+                    (HsInfixApp parseWrapped altOp parseUnwrapped)
+            where
+              bndNameStr = "parse" <> oneofType
+              bndName    = HsVar (unqual_ bndNameStr)
+              parseWrapped = HsParen $
+                HsInfixApp (HsParen (HsInfixApp lambdaVar parseJSONPBOp oneofTyLit))
+                                    bindOp
+                                    (apply (HsVar (jsonpbName "withObject")) [ oneofTyLit , bndName ])
+
+              parseUnwrapped = HsParen (HsApp bndName lambdaVar)
+
+          -- E.g., for message Something{ oneof name_or_id { string name = _; int32 someid = _; } }:
+          -- parseNameOrId =
+          --   fmap Hs.msum
+          --     (Hs.sequence
+          --        [ fmap (Just . SomethingPickOneName) <$> (`HsJSONPB.parseField` "name")
```

Comment:
Same comment here: two levels of `fmap` makes this hard for me to follow what the generated code does

Reply:
Ok, I'll add the parameter back in, at the cost of some generated binder verbosity

```
parseNameOrId parseNameOrId_object = 
  Hs.msum
    [ Just . SomethingPickOneName   <$> (HsJSONPB.parseField parseNameOrId_object "name")
    , Just . SomethingPickOneSomeid <$> (HsJSONPB.parseField parseNameOrId_object "someid")		
    , pure Nothing		
   ]
```